src/package/packageInstall.ts

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright (c) 2022, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import { Connection, Lifecycle, Logger, Messages, PollingClient, SfError, StatusResult } from '@salesforce/core';
+import { isString, isNumber, Optional } from '@salesforce/ts-types';
+import { QueryResult } from 'jsforce';
+import { Duration } from '@salesforce/kit';
+import { PackagingSObjects } from '../interfaces';
+import {
+  isErrorFromSPVQueryRestriction,
+  isErrorPackageNotAvailable,
+  getPackageTypeBy04t,
+  escapeInstallationKey,
+} from '../utils/packageUtils';
+import { consts } from '../constants';
+import { PackageInstallOptions, PackageInstallCreateRequest } from '../interfaces/packagingInterfacesAndType';
+
+import SubscriberPackageVersion = PackagingSObjects.SubscriberPackageVersion;
+import PackageInstallRequest = PackagingSObjects.PackageInstallRequest;
+
+Messages.importMessagesDirectory(__dirname);
+const installMsgs = Messages.loadMessages('@salesforce/packaging', 'package-install');
+
+let logger: Logger;
+const getLogger = (): Logger => {
+  if (!logger) {
+    logger = Logger.childFromRoot('installPackage');
+  }
+  return logger;
+};
+
+export async function installPackage(
+  connection: Connection,
+  pkgInstallCreateRequest: PackageInstallCreateRequest,
+  options?: PackageInstallOptions
+): Promise<PackageInstallRequest> {
+  const defaults = {
+    ApexCompileType: 'all',
+    EnableRss: false,
+    NameConflictResolution: 'Block',
+    PackageInstallSource: 'U',
+    SecurityType: 'none',
+    UpgradeType: 'mixed-mode',
+  };
+
+  const request = Object.assign({}, defaults, pkgInstallCreateRequest);
+  if (request.Password) {
+    request.Password = escapeInstallationKey(request.Password);
+  }
+
+  const pkgType = await getPackageTypeBy04t(request.SubscriberPackageVersionKey, connection, request.Password);
+
+  // Only unlocked packages can change the UpgradeType and ApexCompile options from the defaults.
+  if (pkgType !== 'Unlocked') {
+    if (request.UpgradeType !== defaults.UpgradeType) {
+      const msg = installMsgs.getMessage('upgradeTypeOnlyForUnlockedWarning');
+      await Lifecycle.getInstance().emit('PackageInstallRequest:warning', msg);
+      delete request.UpgradeType;
+    }
+    if (request.ApexCompileType !== defaults.ApexCompileType) {
+      const msg = installMsgs.getMessage('apexCompileOnlyForUnlockedWarning');
+      await Lifecycle.getInstance().emit('PackageInstallRequest:warning', msg);
+      delete request.ApexCompileType;
+    }
+  }
+
+  await Lifecycle.getInstance().emit('PackageInstallRequest:presend', request);
+
+  const result = await connection.tooling.create('PackageInstallRequest', request);
```

Comment:
Do we want to catch unhandled exceptions here to provide better messaging or emit a event?

Reply:
I don't know.  This matches the original behavior.  I wouldn't know what errors to catch to provide better messaging.